### PR TITLE
723. Candy Crush

### PR DESCRIPTION
**Pull request by altayhunoglu**
https://github.com/spiralgo/algorithms/issues/262

image

### CODE REVIEW

File: src/main/java/algorithms/curated170/medium/candycrush/CandyCrushFaster.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                            target[row][col] = 0;
+                        }
+                        targetIndex--;
+                    }
+                }
+            }
+        }
+
+        return crush;
+    }
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/candycrush/CandyCrushFaster.java:60]: We can extract this to a method.
```suggestion
if (crush) {
            drop(target);
        }
        return crush;
    }

    private void drop(int[][] target) {
        for (int col = 0; col < target[0].length; col++) {
            int targetIndex = target.length - 1;
            for (int row = target.length - 1; row >= 0; row--) {
                if (target[row][col] != 0) {
                    if (row != targetIndex) {
                        target[targetIndex][col] = target[row][col];
                        target[row][col] = 0;
                    }
                    targetIndex--;
                }
            }
        }
    }
```

**Comment by ErdemT09** [comment]
I don't think we need the old solution besides its use of recursion. 
Here's the new solution with recursion used:
```
public int[][] candyCrush(int[][] board) {
        int[][] target = new int[board.length][board[0].length];
        if(crush(board, target))
        {
            return candyCrush(target);
        }
        return board;
    }
```

**Comment by altayhunoglu** [comment]
> I don't think we need the old solution besides its use of recursion. 
> 
> Here's the new solution with recursion used:
> 
> ```
> 
> public int[][] candyCrush(int[][] board) {
> 
>         int[][] target = new int[board.length][board[0].length];
> 
>         if(crush(board, target))
> 
>         {
> 
>             return candyCrush(target);
> 
>         }
> 
>         return board;
> 
>     }
> 
> ```

We can simply remove the older.

**Comment by altayhunoglu** [comment]
Oke. We have applied your modularity suggestion.

**Comment by altayhunoglu** [comment]
> Don't we delete the older solution and apply its recursiveness to the newer one?

I am sorry that I recognized you emphasized on the recursive part yet.

Sure, I will try to do it.

**Comment by altayhunoglu** [comment]
> Don't we delete the older solution and apply its recursiveness to the newer one?

Oke, it is done.

File: src/main/java/algorithms/curated170/medium/candycrush/CandyCrush.java [hunk 2]
```diff
@@ -12,132 +10,79 @@ public static void main(String[] args) {
...
+    public int[][] candyCrush(int[][] board) {
+        int[][] target = new int[board.length][board[0].length];
 
-                            boardCopy[i][low] = board[i][low];
-                            low++;
+        if (crush(board, target)) {
+            int[][] temp = board;
+            board = target;
+            target = temp;
+            return candyCrush(board);
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/candycrush/CandyCrush.java:27]: Why do we assign  values to temp and target here when we are not using them again?

Comment (ErdemT09) [reply src/main/java/algorithms/curated170/medium/candycrush/CandyCrush.java:27]: We can just input target to the method.

**Comment by altayhunoglu** [comment]
You are right. It is just a mistake.

**Comment by ErdemT09** [comment]
Smart refactoring for the last one.
For this question, it kind of reminded me the concept of "game loop" from game development:
image

**Review by ErdemT09 — APPROVED** [review]

**Comment by altayhunoglu** [comment]
Thanks for the information and attention Erdem.

**Pull request merged**